src/goto-instrument/function_stubs.cpp

```diff
@@ -0,0 +1,276 @@
+  // function body
+  if(requires.is_not_nil())
+    goto_function.body.add(
+      goto_programt::make_assertion(requires, requires.source_location()));
+
+  // Get the stub implementation function symbol. The stub
+  // implementation function is called from the stub and can be
+  // implemented by the user to give the stub some custom specific
+  // behaviour. If the stub implementation function is left without a
+  // body, CBMC by default treats it as returning a non-deterministic
+  // return value.
+  const symbolt &stub_implementation_symbol =
+    symbol_table.lookup_ref(stub_name_of_function(function_id));
+
+  // Make the function call to the implementation function
+  code_function_callt internal_implementation_call(
+    stub_implementation_symbol.symbol_expr());
+
+  // Declare a variable to store the return value from the internal call
+  symbol_exprt return_value_symbol = get_fresh_aux_symbol(
+                                       type.return_type(),
+                                       id2string(function_id) + "::ret_val",
+                                       "ret_val",
+                                       source_locationt(),
+                                       stub_implementation_symbol.mode,
+                                       symbol_table)
+                                       .symbol_expr();
+
+  replace_symbolt replace_return_value_symbol;
+
+  if(type.return_type() != empty_typet())
+  {
+    // Declare the return variable and assign to it the return value
+    // of the call
+    goto_function.body.add(goto_programt::make_decl(return_value_symbol));
+    internal_implementation_call.lhs() = return_value_symbol;
+
+    // Replace the return_value with r
+    symbol_exprt ret_val(CPROVER_PREFIX "return_value", type.return_type());
+    replace_return_value_symbol.insert(ret_val, return_value_symbol);
+  }
+
+  // Add the parameters as arguments to the call
+  for(const auto &parameter_identifier : type.parameter_identifiers())
+  {
+    const symbolt &parameter_symbol = ns.lookup(parameter_identifier);
+    internal_implementation_call.arguments().push_back(
+      parameter_symbol.symbol_expr());
+  }
+
+  goto_function.body.add(goto_programt::make_function_call(
+    code_function_callt(internal_implementation_call)));
+  // Maybe: Add location/comments maybe?
+
+  // Add the postcondition assumption
+  if(ensures.is_not_nil())
+  {
+    exprt ensures_replaced = ensures;
+    replace_return_value_symbol(ensures_replaced);
+    goto_function.body.add(goto_programt::make_assumption(
+      ensures_replaced, ensures.source_location()));
+  }
+
+  if(type.return_type() != empty_typet())
+  {
+    code_returnt return_expr(return_value_symbol);
+    goto_function.body.add(goto_programt::make_return(return_expr));
+  }
+
+  // Append the end of function instruction
+  goto_function.body.add(
+    goto_programt::make_end_function(ensures.source_location()));
+}
+
+bool function_stubst::operator()()
```

Comment:
`function_stubst` can only stub one function, right? It's also an operator, so I'd suggest calling it `stub_functiont`, and the corresponding free function `stub_function` rather than `function_stubs`, since again it only acts on one function

Reply:
Thanks for the comment. I agree with you. The naming is not very clear. I changed the name of the module as well as of the function to `stub_function`.